src/recursion_guard.rs

```diff
@@ -95,29 +95,26 @@ impl<T: Eq + Hash + Clone> SmallContainer<T> {
     pub fn contains_or_insert(&mut self, v: T) -> Option<usize> {
         match self {
             Self::Array(array) => {
-                let mut first_slot: Option<usize> = None;
-                for (index, op_value) in array.iter().enumerate() {
+                for (index, op_value) in array.iter_mut().enumerate() {
                     if let Some(existing) = op_value {
                         if existing == &v {
                             return None;
                         }
                     } else {
-                        first_slot = first_slot.or(Some(index));
+                        *op_value = Some(v);
```

Comment:
i don't think this is safe because there's a case where an earlier item has been removed from the array, but later values are still set.

E.g.

```py
array = [123, 456, None, 789]
```
If you're looking up `789` it wouldn't be found in this case.

I had this before, but changed it to be safer.

Reply:
I guess if we can be sure that new items are always added to the array (and removed) inside parent calls, we can be sure that there will never be `None` caps in the array.

@davidhewitt if you're confident of that, we can go ahead with this change. Worst case we just fallback on the depth guard.

Reply:
Pushed another commit which should work more like a stack, and we'll panic if the code is wrong.